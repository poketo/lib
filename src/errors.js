// @flow

type ErrorCode =
  | 'INVALID_URL'
  | 'UNSUPPORTED_SITE'
  | 'UNSUPPORTED_SITE_REQUEST'
  | 'TIMEOUT'
  | 'HTTP_ERROR';

class PoketoError extends Error {
  code: ErrorCode;
  constructor(errorCode: ErrorCode, message: string) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.code = errorCode;
  }
}

class InvalidUrlError extends PoketoError {
  constructor(url: string) {
    super('INVALID_URL', `Unable to parse url '${url}'`);
  }
}

class HTTPError extends PoketoError {
  statusCode: number;
  url: string;

  constructor(statusCode: number, message: string, url: string) {
    super('HTTP_ERROR', message);
    this.statusCode = statusCode;
    this.url = url;
  }
}

class NotFoundError extends HTTPError {
  constructor(url: string) {
    super(404, 'Not Found', url);
  }
}

class TimeoutError extends PoketoError {
  constructor(message: string, url: string) {
    super('TIMEOUT', message);
  }
}

class UnsupportedSiteError extends PoketoError {
  constructor(site: string) {
    super('UNSUPPORTED_SITE', `Site at '${site}' is not supported`);
  }
}

class UnsupportedSiteRequestError extends PoketoError {
  constructor(siteName: string, operationName: string) {
    super(
      'UNSUPPORTED_SITE_REQUEST',
      `${siteName} does not support ${operationName}`,
    );
  }
}

export default {
  PoketoError,
  HTTPError,
  InvalidUrlError,
  NotFoundError,
  TimeoutError,
  UnsupportedSiteError,
  UnsupportedSiteRequestError,
};
