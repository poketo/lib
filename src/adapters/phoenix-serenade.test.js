import site from './phoenix-serenade';
import errors from '../errors';

describe('PhoenixSerenadeAdapter', () => {
  const server = new AdapterVcrServer(site);

  beforeAll(async () => {
    await server.listenAndMock(57166);
  });

  afterAll(() => {
    server.close();
  });

  describe('getSeries', () => {
    it('returns a metadata object', async () => {
      const { chapters, ...metadata } = await site.getSeries(
        'kusuriya-no-hitorigoto',
      );

      expect(metadata).toMatchSnapshot();

      const chapterNumbersToTest = ['7', '11'];
      const chaptersToTest = chapters.filter(chapter =>
        chapterNumbersToTest.includes(chapter.chapterNumber),
      );
      expect(chaptersToTest).toMatchSnapshot();
    });
  });
});
