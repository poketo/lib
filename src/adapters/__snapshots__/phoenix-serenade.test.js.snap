// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoenixSerenade getSeries returns a metadata object 1`] = `
Object {
  "author": "HYUUGA Natsu, NANAO Ikki, Nekokurage",
  "coverImageUrl": undefined,
  "description": "In the imperial court, a young woman is put into servitude, Maomao. The tale is just beginning for the woman doctor/pharmacist from the red-light district, as rumors circulate about the emperor's children's lives being short-lived. Her curious nature and thirst for knowledge pushes her to action. To satisfy her curiosity, this young doctor/pharmacist will investigate the origin! What is shifting in the imperial court..!?

http://www.jp.square-enix.com/magazine/biggangan/introduction/kusuriya/",
  "id": "phoenix-serenade:kusuriya-no-hitorigoto",
  "site": Object {
    "id": "phoenix-serenade",
    "name": "Phoenix Serenade",
  },
  "slug": "kusuriya-no-hitorigoto",
  "status": "UNKNOWN",
  "supportsReading": true,
  "title": "Kusuriya no Hitorigoto",
  "updatedAt": Any<Number>,
  "url": "http://localhost:57166/series/kusuriya-no-hitorigoto",
}
`;

exports[`PhoenixSerenade getSeries returns a metadata object 2`] = `
Array [
  Object {
    "chapterNumber": "11",
    "createdAt": 1526538824,
    "id": "phoenix-serenade:kusuriya-no-hitorigoto:en/0/11",
    "order": 5,
    "slug": "en/0/11",
    "title": undefined,
    "url": "http://localhost:57166/read/kusuriya-no-hitorigoto/en/0/11/page/1",
    "volumeNumber": "0",
  },
  Object {
    "chapterNumber": "7",
    "createdAt": 1513954489,
    "id": "phoenix-serenade:kusuriya-no-hitorigoto:en/0/7",
    "order": 1,
    "slug": "en/0/7",
    "title": undefined,
    "url": "http://localhost:57166/read/kusuriya-no-hitorigoto/en/0/7/page/1",
    "volumeNumber": "0",
  },
]
`;
